hs_collection_resource/management/commands/fix_collection_version_missing_ispartof.py

```diff
@@ -32,4 +34,35 @@ def handle(self, *args, **options):
                         res.metadata.create_element('relation',
                                                     type=RelationTypes.isPartOf,
                                                     value=collection.get_citation())
+        print('-' * 100)
+        print("check that all collections include hasPart metadata for contained resources")
+        # for rel in Relation.objects.filter(type=RelationTypes.isPartOf).all():
+        #     col = rel.metadata.resource
+        #     print(f"Checking collection {col}, https://beta.hydroshare.org/resource/{col.short_id}")
+        #     print("This collection is pointed to by a res with isPartOf metadata. Checking that col contains hasPart metadata")
+        #     haspart_relations = col.metadata.relations.filter(type=RelationTypes.hasPart).all()
+        #     print(haspart_relations)
+        #     # Need a way to check that hasPart relations contains the resources with isPart metadata
+
+        for res in BaseResource.object.all():
```

Comment:
I think we can avoid this for loop block and set the hasPart relation for the collection within the loop block for collection above.

Reply:
I don't think we can. Our goal is twofold:
1. Check that all resources in collections have isPartOf metadata (first loop)
The reason we can't just set the hasPart in the first loop is because if the hasPart metadata is missing on the collection, that resource isn't returned when we do `collection.resources.all()` line 27

3. Check that all collections include hasPart metadata for contained resources (second loop)
I think this has to be done separately, iterating all resources that have isPartOf metadata and ensuring that the corresponding collection includes the hasPart metadata

Reply:
```
for collection in CollectionResource.objects.all():
      # code to fix the isPartOf - you already have it

     # code to fix missing hasPart in CollectionResource
     for res in collection.resources.all():
           res_citation = res.get_citation()
           if not collection.metadata.relations.filter(type=RelationeTypes.hasPart, value=res_citation).exists():
              collection.metadata.create_element('relation',
                                                    type=RelationTypes.hasPart,
                                                    value=res_citation)
```

Reply:
thanks @pkdash! Maybe I'm wrong but I think this still has the problem that
`for res in collection.resources.all():` will only return those resources that already have `hasPart` metadata in the collection. So the resource that needs to be added to the collection with `hasPart` would not be included in your `for res in collection.resources.all()` loop

Let me give an example. Here is the resource Dave wanted to add:
https://www.hydroshare.org/resource/45b43d72928048a8bc10a009d932f769/ 
```
>>> res = BaseResource.objects.filter(short_id='45b43d72928048a8bc10a009d932f769').all()
>>> rel = res.first().metadata.relations.all()
>>> for re in rel:
...     print(re)
isPartOf Tarboton, D. (2023). Collection of Great Salt Lake Data, HydroShare, http://www.hydroshare.org/resource/b6c4fcad40c64c4cb4dd7d4a25d0db6e
isVersionOf Tarboton, D. (2017). Great Salt Lake Level and Volume, HydroShare, http://www.hydroshare.org/resource/b26090299ec947c692d4ee4651815579
references Tarboton, D. (2017). Great Salt Lake Area Volume Data, HydroShare, http://www.hydroshare.org/resource/89125e9a3af544eab2479b7a974100ba
```
Okay we see that the resource contains isPartOf metadata for this collection:
http://www.hydroshare.org/resource/b6c4fcad40c64c4cb4dd7d4a25d0db6e

Now let's check the collection
```
>>> from hs_core.models import BaseResource, Relation
>>> col = BaseResource.objects.filter(short_id='b6c4fcad40c64c4cb4dd7d4a25d0db6e').all()
>>> rels = col.first().metadata.relations.all()
>>> for rel in rels:
...     print(rel)
...
hasPart Tarboton, D. (2017). Great Salt Lake Area Volume Data, HydroShare, http://www.hydroshare.org/resource/89125e9a3af544eab2479b7a974100ba
hasPart Tarboton, D. (2017). Great Salt Lake Bathymetry, HydroShare, http://www.hydroshare.org/resource/582060f00f6b443bb26e896426d9f62a
hasPart Tarboton, D. (2022). Great Salt Lake Basin Digital Elevation Model, HydroShare, http://www.hydroshare.org/resource/e660640a7b084794aa2d70dc77cfa67b
hasPart Merck, M., D. Tarboton (2023). Great Salt Lake Mineral Extraction Water Withdrawals, HydroShare, http://www.hydroshare.org/resource/46b98820b7214850aadf6fa60c9bc1d1
hasPart Tarboton, D. (2023). Great Salt Lake Basin Watershed Boundaries Shapefile, HydroShare, http://www.hydroshare.org/resource/2796b46fa75140be821032846d6140c0
```
^^^ we see that the collection DOESNT have the proper `hasPart` metadata element pointing back at the resource.
And if we run
```
>>> resources_in_col = col.first().resources.all()
>>> for r in resources_in_col:
...     print(r)
...     print(r.short_id)
...
Great Salt Lake Area Volume Data
89125e9a3af544eab2479b7a974100ba
Great Salt Lake Basin Digital Elevation Model
e660640a7b084794aa2d70dc77cfa67b
Great Salt Lake Bathymetry
582060f00f6b443bb26e896426d9f62a
Great Salt Lake Mineral Extraction Water Withdrawals
46b98820b7214850aadf6fa60c9bc1d1
Great Salt Lake Basin Watershed Boundaries Shapefile
2796b46fa75140be821032846d6140c0
```
We see that https://www.hydroshare.org/resource/45b43d72928048a8bc10a009d932f769/  is not included in `col.first().resources.all()`

So if I run your code suggestion on this collection:
```
>>> collection = CollectionResource.objects.filter(short_id='b6c4fcad40c64c4cb4dd7d4a25d0db6e').first()
>>> for res in collection.resources.all():
...     print(f"Checking res {res}")
...     res_citation = res.get_citation()
...     if not collection.metadata.relations.filter(type=RelationTypes.hasPart, value=res_citation).exists():
...         print(f"need to make hasPart with value={res_citation}")
...
Checking res Great Salt Lake Area Volume Data
Checking res Great Salt Lake Basin Digital Elevation Model
Checking res Great Salt Lake Bathymetry
Checking res Great Salt Lake Mineral Extraction Water Withdrawals
Checking res Great Salt Lake Basin Watershed Boundaries Shapefile
```
We can see that it fails to detect the missing hasPart relation

Reply:
@devincowan if this resource https://www.hydroshare.org/resource/45b43d72928048a8bc10a009d932f769/ has the metadata isPartOf pointing to a collection and we can find that resource in col.resources.all() then the relationship is broken at the DB level in addition to the metadata level (Relation meta).

Reply:
you said "...and we can find that resource in col.resources.all()..."
did you mean "...and we **CAN'T** find that resource in col.resources.all()..." ???

Reply:
@devincowan The relationship between a collection resource and resource that it contains at the db level is the real relation. We have to fix the metadata accordingly.